src/main/java/frc/robot/vision/sources/limelights/LimeLightSource.java

```diff
@@ -0,0 +1,167 @@
+import frc.constants.VisionConstants;
+import frc.robot.vision.data.AprilTagVisionData;
+import frc.robot.vision.sources.GyroRequiringVisionSource;
+import frc.utils.Conversions;
+import frc.utils.Filter;
+import frc.utils.alerts.Alert;
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+import frc.utils.time.TimeUtils;
+
+import java.util.Optional;
+
+public class LimeLightSource implements GyroRequiringVisionSource {
+
+	private final NetworkTableEntry robotPoseEntryBotPose2;
+	private final NetworkTableEntry robotPoseEntryBotPose1;
+	private final NetworkTableEntry aprilTagIdEntry;
+	private final NetworkTableEntry aprilTagPoseEntry;
+	private final NetworkTableEntry robotOrientationEntry;
+	private final String name;
+	private Filter<AprilTagVisionData> filter;
+	private double[] robotPoseArray;
+	private double[] aprilTagPoseArray;
+	private Rotation2d robotHeading;
+	private GyroAngleValues gyroAngleValues;
+	private boolean useGyroForPoseEstimating;
+
+	public LimeLightSource(String name, String parentLogPath, Filter<AprilTagVisionData> filter) {
+		String logPath = parentLogPath + name + "/";
+		this.name = name;
+		this.filter = filter;
+		this.robotPoseEntryBotPose2 = getLimelightNetworkTableEntry("botpose_orb_wpiblue");
+		this.robotPoseEntryBotPose1 = getLimelightNetworkTableEntry("botpose_wpiblue");
+		this.aprilTagPoseEntry = getLimelightNetworkTableEntry("targetpose_cameraspace");
+		this.aprilTagIdEntry = getLimelightNetworkTableEntry("tid");
+		this.robotOrientationEntry = getLimelightNetworkTableEntry("robot_orientation_set");
+		this.gyroAngleValues = new GyroAngleValues(Rotation2d.fromDegrees(0), 0, Rotation2d.fromDegrees(0), 0, Rotation2d.fromDegrees(0), 0);
+		this.useGyroForPoseEstimating = true;
+
+		AlertManager.addAlert(
+			new PeriodicAlert(
+				Alert.AlertType.WARNING,
+				logPath + "DisconnectedAt",
+				() -> getLimelightNetworkTableEntry("tv").getInteger(-1) == -1
+			)
+		);
+	}
+
+	@Override
+	public void update() {
+		robotOrientationEntry.setDoubleArray(
+			new double[] {
+				gyroAngleValues.yaw().getDegrees(),
+				gyroAngleValues.yawRate(),
+				gyroAngleValues.pitch().getDegrees(),
+				gyroAngleValues.pitchRate(),
+				gyroAngleValues.roll().getDegrees(),
+				gyroAngleValues.rollRate()}
+		);
+		robotPoseArray = (useGyroForPoseEstimating ? robotPoseEntryBotPose2 : robotPoseEntryBotPose1)
+			.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		aprilTagPoseArray = aprilTagPoseEntry.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		double[] robotPoseWithoutGyroInput = robotPoseEntryBotPose1.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		robotHeading = Rotation2d.fromDegrees(robotPoseWithoutGyroInput[LimelightEntryValue.YAW_ANGLE.getIndex()]);
+	}
+
+	private Optional<Pair<Pose3d, Double>> getUpdatedPose3DEstimation() {
+		int id = (int) aprilTagIdEntry.getInteger(VisionConstants.NO_APRILTAG_ID);
```

Comment:
add the safety comment here

Reply:
wdym? 
this casting is safe since it's the only thing that can be in this networktable entry. It returns long and we cast it into an integer, but that's just how the API is working.